### PR TITLE
Add selected row state

### LINKED ISSUE #156: Outbound shipment: More row selection fun!
- If you select a row, then sort, the selection is removed. This is because the `data` changes (comes from the server) and is the intended behaviour.
- If you add `autoResetSelectedRows: false` - then the rows are kept selected, however they're selected by index. This means if you select row 2, then change the sort order, row 2 continues to be sorted.

So...

- Is this OK? I would think that we should keep the rows selected 🤔 
- I also had a thought around selecting on one page then going to the next page - obviously we lose the data of the next page, but wondered if the user expectation is that you can select all rows on a page, then go to the next, select all the rows, and perform an action on both pages

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #156 

This is what I ended up with. Not 100% happy and open to suggestions or scraping. Confused myself. Why do I do this!

### CODE REVIEW

File: packages/transactions/src/OutboundShipment/DetailView/DetailView.tsx [hunk 1]
```diff
@@ -187,3 +189,11 @@ export const OutboundShipmentDetailView: FC = () => {
...
   ) : null;
 };
+
+export const OutboundShipmentDetailView: FC = () => {
+  return (
+    <TableProvider createStore={createTableStore}>
+      <OutboundShipmentDetailViewComponent />
+    </TableProvider>
+  );
+};
```
Comment (joshxg) [inline packages/transactions/src/OutboundShipment/DetailView/DetailView.tsx:199]: Now I've added the `useTableStore` into the `RemoteDataTable` the context becomes mandatory` 🤔

File: packages/transactions/src/OutboundShipment/ListView/ListView.tsx [hunk 2]
```diff
@@ -23,16 +22,65 @@ import {
...
+  createTableStore,
+  getCheckboxSelectionColumn,
+  ColumnDefinition,
+  useTableStore,
 } from '@openmsupply-client/common';
 
 import { OutboundShipmentListViewApi } from '../../api';
 
-export const OutboundShipmentListView: FC = () => {
+const ListViewToolBar: FC<{
```
Comment (joshxg) [inline packages/transactions/src/OutboundShipment/ListView/ListView.tsx:34]: Just shifting it out so avoid massive re-rendering of the listview when a select happens which then slices the data again aaand..

Looking to extract this out anyway, but a bit nicer soon

File: packages/common/src/ui/layout/tables/context/TableContext.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        // If this row is being toggled on, add one, otherwise reduce the number
+        // of rows selected.
+        const newNumberSelected = numberSelected + (isSelected ? 1 : -1);
+
+        return {
+          ...state,
+          numberSelected: newNumberSelected,
+          rowState: {
+            ...state.rowState,
+            [id]: { ...state.rowState[id], isSelected },
```
Comment (mark-prins) [inline packages/common/src/ui/layout/tables/context/TableContext.ts:86]: I'm not sure the point of spreading `state.rowState[id]` as `isSelected` is the only prop

File: packages/common/src/ui/layout/tables/types.ts [hunk 4]
```diff
@@ -13,7 +13,7 @@ export enum ColumnFormat {
 }
 
 export interface ColumnDefinition<T> {
-  label: LocaleKey;
+  label?: LocaleKey;
```
Comment (mark-prins) [inline packages/common/src/ui/layout/tables/types.ts:16]: hm? what did I miss?

File: packages/transactions/src/OutboundShipment/DetailView/DetailView.tsx [hunk 5]
```diff
@@ -80,7 +82,7 @@ const useDraftOutbound = (id: string) => {
   return { draft, setDraft, save };
 };
 
-export const OutboundShipmentDetailView: FC = () => {
+export const OutboundShipmentDetailViewComponent: FC = () => {
```
Comment (mark-prins) [inline packages/transactions/src/OutboundShipment/DetailView/DetailView.tsx:85]: doesn't need to exported

File: packages/transactions/src/OutboundShipment/DetailView/DetailView.tsx (hunk 1 above)
Comment (mark-prins) [reply packages/transactions/src/OutboundShipment/DetailView/DetailView.tsx:199]: have you tried with two tables on a page? that would work, right? having a different provider wrapped around each.

I'm concerned that any component which uses a data table has to remember to wrap it in a provider, though I think I'm worrying about nothing and it is probably fine

**Review by mark-prins — APPROVED** [review]
👍  I think its great - took me a little time to understand what the test descriptions were all about, in the end it's all clear enough.

File: packages/common/src/ui/layout/tables/types.ts (hunk 4 above)
Comment (joshxg) [reply packages/common/src/ui/layout/tables/types.ts:16]: The checkbox column doesn't need the label

File: packages/common/src/ui/layout/tables/context/TableContext.ts (hunk 3 above)
Comment (joshxg) [reply packages/common/src/ui/layout/tables/context/TableContext.ts:86]: Just so if we add, say, a `isDisabled` this still works, I guess!

**Comment by joshxg** [comment]
two pages: Yeah works, but need to wrap each one in it's own context.

Context concern in general: I have the same concern. Technically the wrapper is not needed now since extracting the `useTableStore` to the `ListViewToolbar` component. But the table in general does need to be wrapped in the context. I think I convinced myself by material-ui moving toward a `TabContext` if you want to use tabs and I likened this to having a `toolbar` (the dropdown which needs to know which rows are selected) and the table.

Another factor that i'm a little unsure about is why the row select state is in context, but the sort by state isn't? Or the pagination state? 🤔

**Pull request merged**